### PR TITLE
Add MLFLOW_FLATTEN_PARAMS support in MLflowCallback

### PR DESCRIPTION
**Pull request by orieg**
# What does this PR do?

This PR add support for the environment variable `MLFLOW_FLATTEN_PARAMS`.

When first level parameters hold a dictionary value, it will be logged to MLflow as a string.  Currently, it will skip that parameter when the string exceed 250 characters. This is especially true with the task_specific_params which can end up being a long string.

Current warning message look like this:

> Trainer is attempting to log a value of "{'summarization': {'length_penalty': 1.0, 'max_length': 128, 'min_length': 12, 'num_beams': 4}, 'summarization_cnn': {'length_penalty': 2.0, 'max_length': 142, 'min_length': 56, 'num_beams': 4}, 'summarization_xsum': {'length_penalty': 1.0, 'max_length': 62, 'min_length': 11, 'num_beams': 6}}" for key "task_specific_params" as a parameter. MLflow's log_param() only accepts values no longer than 250 characters so we dropped this attribute.

With this PR, the warning message is updated to:
> Trainer is attempting to log a value of "{'summarization': {'length_penalty': 1.0, 'max_length': 128, 'min_length': 12, 'num_beams': 4}, 'summarization_cnn': {'length_penalty': 2.0, 'max_length': 142, 'min_length': 56, 'num_beams': 4}, 'summarization_xsum': {'length_penalty': 1.0, 'max_length': 62, 'min_length': 11, 'num_beams': 6}}" for key "task_specific_params" as a parameter. MLflow's log_param() only accepts values no longer than 250 characters so we dropped this attribute. You can use `MLFLOW_FLATTEN_PARAMS` environment variable to flatten the parameters and avoid this message.

When a user set the env variable with `os.environ['MLFLOW_FLATTEN_PARAMS'] = "True"`, the parameters will be properly sent to MLflow and logged as such:

```
task_specific_params.summarization.length_penalty	1.0
task_specific_params.summarization.max_length	128
task_specific_params.summarization.min_length	12
task_specific_params.summarization.num_beams	4
task_specific_params.summarization_cnn.length_penalty	2.0
task_specific_params.summarization_cnn.max_length	142
task_specific_params.summarization_cnn.min_length	56
task_specific_params.summarization_cnn.num_beams	4
task_specific_params.summarization_xsum.length_penalty	1.0
task_specific_params.summarization_xsum.max_length	62
task_specific_params.summarization_xsum.min_length	11
task_specific_params.summarization_xsum.num_beams	6
```

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@sgugger I added a flatten_dict function in .utils/py_utils.py as it didn't seem right to add this in integrations.py.

### CODE REVIEW

**Comment by orieg** [comment]
@sgugger the script `utils/check_inits.py` used for CI has a bug. Line 252:

```
submodule = short_path.replace(os.path.sep, ".").replace(".py", "")
```

This would lead to files starting by "py" to be improperly named. For example, `utils/py_utils.py` is interpreted as `utils_utils`.

A better replace() usage may be:
```
submodule = short_path.replace(".py", "").replace(os.path.sep, ".")
```

File: src/transformers/utils/py_utils.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Python utilities"""
+
+from collections.abc import MutableMapping
+
+
+def flatten_dict(d: MutableMapping, parent_key: str = "", delimiter: str = "."):
```
Comment (sgugger) [inline src/transformers/utils/py_utils.py:20]: This should go in utils.generic instead, not need to create a new file for this IMO

**Review by sgugger — COMMENTED** [review]
Thanks a lot for your PR! Could you move the new `flatten_dict` function to the `utils.generic` module? Also could you add a test for it?

File: src/transformers/utils/py_utils.py (hunk 1 above)
Comment (orieg) [reply src/transformers/utils/py_utils.py:20]: Will do

File: tests/utils/test_generics.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+            "task_specific_params.summarization_xsum.length_penalty": 1.0,
+            "task_specific_params.summarization_xsum.max_length": 62,
+            "task_specific_params.summarization_xsum.min_length": 11,
+            "task_specific_params.summarization_xsum.num_beams": 6,
+        }
+
+        self.assertEqual(
+            flatten_dict(input_dict),
+            expected_dict,
+        )
```
Comment (sgugger) [inline tests/utils/test_generics.py:48]:
```suggestion
        self.assertEqual(flatten_dict(input_dict), expected_dict)
```

File: utils/check_inits.py [hunk 3]
```diff
@@ -249,7 +249,7 @@ def get_transformers_submodules():
             if fname == "__init__.py":
                 continue
             short_path = str((Path(path) / fname).relative_to(PATH_TO_TRANSFORMERS))
-            submodule = short_path.replace(os.path.sep, ".").replace(".py", "")
+            submodule = short_path.replace(".py", "").replace(os.path.sep, ".")
```
Comment (sgugger) [inline utils/check_inits.py:252]: Out of curiosity, why is this change necessary?

**Review by sgugger — APPROVED** [review]
Thanks for addressing the comments!

File: utils/check_inits.py (hunk 3 above)
Comment (orieg) [reply utils/check_inits.py:252]: Not anymore. That was when I had utils/py_utils.py but since it's moved to utils/generic.py, it's not a problem. It may become a problem later for other folks who name a file starting with /py tho.

Comment (sgugger) [reply utils/check_inits.py:252]: Good to keep then!

File: utils/tests_fetcher.py [hunk 4]
```diff
@@ -268,7 +268,7 @@ def create_reverse_dependency_map():
     "feature_extraction_sequence_utils.py": "test_sequence_feature_extraction_common.py",
     "feature_extraction_utils.py": "test_feature_extraction_common.py",
     "file_utils.py": ["utils/test_file_utils.py", "utils/test_model_output.py"],
-    "utils/generic.py": ["utils/test_file_utils.py", "utils/test_model_output.py"],
+    "utils/generic.py": ["utils/test_file_utils.py", "utils/test_model_output.py", "utils/test_generic.py"],
```
Comment (sgugger) [inline utils/tests_fetcher.py:271]: Nice catch!

**Comment by sgugger** [comment]
Looks like you messed the rebase a little bit and there are now commits in this PR that shouldn't be here.

**Comment by sgugger** [comment]
Thanks again!
